File: YeChanSong_코딩테스트_연습/그래프/순위.py
def dfsall(n):
    global dfshist, visited
    for i in range(n):
        visited = list(False for i in range(n))
        cntr = dfs(i,n)
        dfshist[i] = cntr
    
def dfs(idx,n):
    global adj, visited, counting
    visited[idx]=True
    cntr = 1
    for i in adj[idx]:
        if visited[i]==True:
            continue
        cntr += dfs(i,n)
    
    counting[idx] +=1
    return cntr
    
def solution(n, results):
    # dfshist: dfs결과 방문한 노드 수
    global adj, dfshist, counting
    answer = 0
    adj = list([] for i in range(n))
    dfshist = list(-1 for i in range(n))
    counting = list(0 for i in range(n))
    for d,s in results:
        adj[s-1].append(d-1)
    dfsall(n)
    for i in range(n):
        if dfshist[i]+counting[i] == n+1:
            answer+=1
    return answer

File: YeChanSong_코딩테스트_연습/그래프/순위.java
import java.util.*;
import java.io.*;

class Solution {
    private static int[] visitCount;
    private static int[] eachBfsResult;
    private static boolean[] visited;
    private static ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
    private static int totalVisitCount = 1;

    public int solution(int n, int[][] results) {
        int answer = 0;

        for(int i=0;i<n+1;i++) adj.add(new ArrayList<>());

        Arrays.stream(results).forEach( (result) -> {
            adj.get(result[0]).add(result[1]);
        });

        visitCount = new int[n+1];
        eachBfsResult = new int[n+1];

        for(int i=1;i<n+1;i++) bfs(n, i);
        for(int i=1;i<n+1;i++) {
            if (visitCount[i] + eachBfsResult[i] == n+1) answer++;
        }

        return answer;
    }

    private void bfs(int n, int start) {
        totalVisitCount = 1;
        Queue<Integer> queue = new ArrayDeque<>();
        visited = new boolean[n+1];
        visited[start] = true;
        visitCount[start]++;
        queue.offer(start);

        while(!queue.isEmpty()) {
            int current = queue.poll();
            adj.get(current).forEach( (next) -> {
                if(!visited[next]) {
                    visited[next] = true;
                    queue.offer(next);
                    totalVisitCount+=1;
                    visitCount[next]++;
                }
            });
        }

        eachBfsResult[start] = totalVisitCount;

    }
}